### DISCUSSION TITLE
Expose Server Name Indication (SNI) within ClientTLS autorization request claims

### DISCUSSION THREAD

**Discussion by charbonnierg** [nats-io/nats-server#4706] [Ideas]
I'm using an auth callout service to authenticate/authorize users on a TLS enabled NATS server.

I would like to know the server name that was used during client TLS handshake within my auth callout service.

Like in this example below, I would like to deny access if the proper server name is not used, even if token is valid:
- first attempt does not have a token: access is denied
- second attempt uses correct server name and correct token: access is granted
- third attempt uses invalid server name and correct token: access is denied
![image](https://github.com/nats-io/nats-server/assets/81654818/e9badef4-f93d-4554-866c-f8719c69b73f)

In order to do that, I had to modify both the [jwt](https://github.com/nats-io/jwt/compare/main...charbonnierg:jwt:auth_claims/server_name) and [nats-server](https://github.com/nats-io/nats-server/compare/main...charbonnierg:nats-server:auth_claims/server_name) repositories and added support for server name in request claims.

I've got two questions:

- Is there any way to do the same thing without modifying jwt and nats-server repositories ?
- If no, are you open to adding the server name in ClientTLS struct ? I could create a PR

**Comment by derekcollison** [comment]
That information is already present.

<img width="644" alt="image" src="https://github.com/nats-io/nats-server/assets/90097/f920f6f0-a2c7-4103-9366-51f83de7f651">

**Comment by charbonnierg** [comment]
Thanks for the reply !

**Comment by charbonnierg** [comment]
I just looked again, but that's not the same thing isn't it ?

I may have host set to `0.0.0.0`. That's the case in my JWT claims:

```json
{
  "aud": "nats-authorization-request",
  "exp": 1698248970,
  "jti": "KUEYEBMA5WXYHIMFWXKYV4GKB234KZPT7XTEQKUDKFB3WJM6IHPQ",
  "iat": 1698248968,
  "iss": "ND2VRW5IA7UIZTAJ4U2C2ZW3ROJSRLENP2CXLEIAWGEATT5XS7SSXTSY",
  "sub": "ACMHTF46SWZIFZMOUKPC7YL2L3EGMEM7NYGBYSNXQBOAHI6XC2VUWDRI",
  "nats": {
    "server_id": {
      "name": "srv-01",
      "host": "0.0.0.0",
      "id": "ND2VRW5IA7UIZTAJ4U2C2ZW3ROJSRLENP2CXLEIAWGEATT5XS7SSXTSY",
      "version": "2.11.0-dev"
    },
    "user_nkey": "UDXT3HLG24IHVVT2U67SHNY24IEK5G2ZA32UH4G3PBHTGRMAWXYOI2NE",
    "client_info": {
      "host": "127.0.0.1",
      "id": 10,
      "user": "TEST",
      "name": "NATS CLI Version 0.0.35",
      "kind": "Client",
      "type": "websocket"
    },
    "connect_opts": {
      "auth_token": "TEST",
      "name": "NATS CLI Version 0.0.35",
      "lang": "go",
      "version": "1.19.0",
      "protocol": 1
    },
    "client_tls": {
      "version": "1.3",
      "cipher": "TLS_AES_128_GCM_SHA256",
      "server_name": "other.local.quara-dev.com"
    },
    "type": "authorization_request",
    "version": 2
  }
}
```

What I want is really the server name used during TLS handshake, not a server name configured on server side. In my case, server accepts connections on several domain names, which are unrelated to server name.

Note: the `server_name` appears in my claims because I use my fork of nats-server

**Reply by derekcollison** [reply]
That server name is from the server that received the connection directly from the client.

Do you have a proxy or LB or TLS terminator in between?

**Reply by charbonnierg** [reply]
No, I have TLS Websocket enabled and I connect from the CLI without any proxy.

Auth service is running within the same process as NATS server so uses an in process connection.

**Reply by derekcollison** [reply]
So that is the server name then.

**Reply by ripienaar** [reply]
It’s not the server name or id. It’s the dns named the client used to connect. 

As it can be anything and is unverified and can even come from /etc/hosts of the client I don’t think we want to expose this here. 

It might lead people to believe it’s a trusted piece of information like most everything else here.

**Reply by charbonnierg** [reply]
What do you mean by that is the server name ?

Regardless of configured NATS server name or listen address, a client connection over TLS must use a "server name", but those are not the same thing.

What I'm referring to as server name is the hostname used during TLS handshake, which is the domain name used in server URL by client.

And a single NATS server may be reachable through different domain names at the same time.

**Reply by derekcollison** [reply]
The server identity is present, meaning info about the actual NATS server that is calling out. We also include the complete TLS chain from the client, which sounds like what you need.

But note, that only will give you insight that the TLS connection succeeded, meaning the hostname they used client side matched to the TLS server cert. If the server cert uses multiple, you will not be able to distinguish which was used client side.

**Reply by charbonnierg** [reply]
Unless I'm mistaken, complete TLS chain is present only when client certificates are used, but not when only server certificate is used. So I cannot rely on TLS chain. And server info is not enough to decide which auth policy should be used.

Also, @ripienaar, could you elaborate why you think that it can be anything ? Since this information comes from the TLS connection state, it means that a valid certificate was found for such hostname. As I understand, it restricts the possible values to values for which server can present a valid certificate for so that's not really "anything".

@derekcollison , maybe what's bugging you is that I'm trying to decide *which account* the authentication request is for within my auth callout. In order words, depending on the domain name used, the auth callout will use a different auth policy, and sign claims for a different account.

**Reply by derekcollison** [reply]
If we have a TLS connection we might be able to extract it from the handshake, I can take a look into that.

**Reply by charbonnierg** [reply]
FWIW I used the server name from [the TLS connection state](https://github.com/charbonnierg/nats-server/blob/b7ca23995072f3e4dcb1c88ba3442eac10974ac1/server/auth_callout.go#L360) and tested it manually, it works great.

Also, thanks for your time and replies, I'm sure you've got many things to do, I really appreciate that you take time to discuss things

**Reply by derekcollison** [reply]
Nice, as you pointed out for this one it has to touch a bunch of libs not just the server, but I am happy to look at PRs if you want to give it a go.

**Reply by charbonnierg** [reply]
I just opened https://github.com/nats-io/jwt/pull/210 which seems to be a pre-requisite for any change in nats-server repo.

I only added the new field to the struct, and did not add any test case ATM, feel free to tell me if anything if required.

**Reply by derekcollison** [reply]
Sounds good, will take a peek.

**Comment by ripienaar** [comment]
Agree @charbonnierg it's not the same.  I am curious what the use case is for verifying this in callout? Not saying its an invalid idea or anything I am just curious :)

**Reply by derekcollison** [reply]
server_id from above contains the info on the server that has the client connection.

**Comment by charbonnierg** [comment]
If it can help understanding the feature request, I have an experimental auth callout plugin system configurable using policies. A policy has:

- a set of matchers: responsible for deciding if policy should apply
- a handler: responsible for either returning an error or user claims

And I'm very interested in having a matcher based on tls server name, just like in the example below:

- For both domains, users should be authorized using oauth2 session ticket provided as token, and if an oauth2 session exists, user claims are signed for the target account.

- For the test app, a specific azure ad enteprise application is used
- For the staging app, a different azure ad enterprise application is used

```
default deny
policy app-users {
	match {
		tls_server_name staging.myapp.com
	}
	callout oauth2 {
		endpoint az-stag
		account STAG_APP
		template {
			allow_pub "users.notify.{oidc.session.email}.>"
			allow_sub "users.service.{oidc.session.email}.>"
			allow_resp
		}
	}
}
policy test-app-users {
	match {
		tls_server_name test.myapp.com
	}
	handler oauth2 {
		endpoint az-test
		account TEST_APP
		template {
			allow_pub "users.notify.{oidc.session.email}.>"
			allow_sub "users.service.{oidc.session.email}.>"
			allow_resp
		}
	}
}
policy admin {
	match {
		connection_type in_process
		username SYS
	}
	callout allow {
		account SYS
	}
}
policy auth-service {
	match {
		connection_type in_process
		username OAUTH2
	}
	callout allow {
		account OAUTH2
	}
}
```

**Reply by ripienaar** [reply]
But surely stage and test isnt the same underlying NATS Server, so the server name would be different?

**Reply by charbonnierg** [reply]
In my case it is the same server. It may be a bad example, but the goal here is to choose which authorization strategy apply according to domain name used at connect.

A more generic example is to authorize connect requests on `customerA.somedomain.com` using an auth policy, and requests on `customerB.somedomain.com` using a different auth policy.

Do you think that is a weird idea ?

**Reply by ripienaar** [reply]
The problem is that nothing prevents customerA from using customerB url - so its not really a security check I think, its more like a soft selector of which rules to apply which seems a bit weird.  It's essentially unverified user supplied data.

**Reply by charbonnierg** [reply]
It's not meant to be a security check, it's meant to match which auth apply and which account is targeted.

I don't really want to have to specify an account through a connect option on user side, and I don't want to map OIDC claims to an account, because a single identity may be granted access to different accounts. What I need is to choose which OIDC client I'm going to use for authentication.

If we look at HTTP, that's a very standard thing to do, you listen to requests for a specific domain name:

- http://nginx.org/en/docs/http/server_names.html
- https://caddyserver.com/docs/caddyfile#the-caddyfile
- https://doc.traefik.io/traefik/routing/routers/#rule

**Reply by charbonnierg** [reply]
And if customerB tries to access customerA URL, he will get an unauthorized access, just like when you visit a private project on github and you don't have access (for example)

**Reply by charbonnierg** [reply]
The users must present cookies which are only valid for a given set of domain names to begin with. So unless user is authenticated AND authorized by the configured OIDC client, it won't be granted access.

You're totally right when you're saying this is a selector, and it's indeed a soft selector because it can be combined with others (such as client infos or connect opts).

I don't agree with the fact that it is unverified user data however, because it is the server name which was used during TLS handshake. So it means that it was a valid server name according to tls config on the server, it's not totally unverified. Also, I want it  to communicate the *intent* of the client, e.g., to which account the client should connect to. The client is free to choose, but may be denied access if OIDC client decides so.

**Reply by ripienaar** [reply]
@charbonnierg may point is this, in your case you correctly understand the purpose of this field and you are using it correctly.  I am not saying you are doing it wrong or using this feature badly. You know this is a unverified information and use it only to choose which strong verification to use - all good.

But this information is user supplied and if a user can see a certificate (they can) they can also see all the valid values for this field and where there are wildcards etc, this is useful information for an attacker to start probing your system and crafting an attack by trying to supply unexpected values or trying to see if maybe a system have reduced controls for some of the valid values or worse there are valid values for which someone forgot to add any control.

Now again, you understand the purpose of this field, but I feel its quite a nuanced situation that many may not understand well.  Someone might assume `lets just allow all connections to dev because they are using our internal hostname` but that is not true because you could put dev hostnames in your /etc/hosts and connect to those names and this field will get this value from anywhere where the connection is possible.  It is inherently untrustable information. Worse its trivially enumarable untrusted value that attackers can easily just try all possible values of or trivially know where they can stuff interesting values into (wildcards)

My concern is adding a specifically untrusted piece of information among otherwise trusted information and this might lead people to make some mistakes - not saying you are making them - I am thinking about the community at large.  That this is a sharp edge that might confuse people or lead them to use this feature incorrectly.

Interested in what @philpennock thinks here

**Reply by charbonnierg** [reply]
@ripienaar I'm no TLS expert and maybe I'm wrong, so I would also be interested in opinions of others.

I'm struggling to understand how SNI can be "a specifically untrusted piece of information". This is an information available within the TLS handshake, which is defined in [RFC 3546](https://datatracker.ietf.org/doc/html/rfc3546) .

It is clearly stated in [this RFC](https://datatracker.ietf.org/doc/html/rfc3546#section-3.1:~:text=A%20server%20that%20receives%20a%20client%20hello%20containing%20the%20%22server_name%22%0A%20%20%20extension%2C%20MAY%20use%20the%20information%20contained%20in%20the%20extension%20to%0A%20%20%20guide%20its%20selection%20of%20an%20appropriate%20certificate%20to%20return%20to%20the%0A%20%20%20client%2C%20and/or%20other%20aspects%20of%20security%20policy.):

>  A server that receives a client hello containing the "server_name"
   extension, MAY use the information contained in the extension to
   guide its selection of an appropriate certificate to return to the
   client, and/or other aspects of security policy. 
   
If server cannot present a certificate for the SNI provided by client during TLS handshake, then an error occurs, and client cannot connect. As such, the sni value cannot be anything, it must be a valid server name according to server tls config as far as I understand.

And in case there is no TLS, there is no SNI to begin with.

**Reply by ripienaar** [reply]
In this case the server is using this piece of user supplied information correctly - its using it to pick the next step where it will do secure validation using a certificate.

My problem with this is not about using the feature correctly (its great when used correctly) it's the ease of using it incorrectly that I am concerned about.  If this conversation highlights anything its that its really hard to get right or even understand - and this makes me weary.

**Reply by charbonnierg** [reply]
I fear that you misunderstood what I'm suggesting (if not sorry for the long message, but it may be useful to others).

It's the NATS server which must present a TLS certificate matching the SNI provided by the user, this has nothing to do with auth callout.

Also it's the NATS server which fails to perform TLS handshake if no certificates are available for user-provided SNI, and auth callout is never called in such case.

One the TLS handshake is complete, NATS server sends a request to the auth callout service. When creating the request claims to be sent to the auth service, the NATS server has access to the TLS connection state (of client connection).

What I'm suggesting is to put SNI used to perform TLS handshake within the request claims.

Sure SNI comes from the user in the first place, but at this point, the server merely indicates which SNI was used for the successful TLS handshake. It is neither untrusted, nor complicated, it's a metadata regarding a TLS handshake which already happened and was successful.

Also, auth callout seems like an advanced usecase, and usage of the client TLS infos seems even more like a niche usecase. I expect users of such features to have prior knowledge of TLS protocol, don't you think so ?

Finally, I don't find arguments regarding misusage very convincing. I mean when writing an auth callout service there are many ways to open security holes. I don't see why having access to the SNI that was provided during connection would lead to more risky situations than having access to connect opts in the first place.

I think that at this point, I gave all the arguments that I had in mind. Again, thanks for your time and the stimulating discussion 🙂

**Reply by ripienaar** [reply]
I understand exactly what you are asking.  The crux of my problem was stated by you:

```
.....matching the SNI provided by the user
```

You are adding a feature that might get people to think this user supplied SNI is something they can trust and its not.  Simple as that. If you support 10 names a user can enumerate them and probe your security because perhaps you used the SNI name incorrectly.  If we make it really easy to use the SNI name incorrectly by not clearly indicating its not to be trusted, users will treat it as trusted.

The server merely indicates what is provided to auth callout, where Joe every day user has to exactly understand all the intricacies discussed here in order to use this provided information correctly.  If they do not understand what the information mean they migth use it incorrectly leading to a security failure in their callout logic.

Contrast that with other information in the certificate etc that is not user provided - its held in a signed certificate and cannot change.

**Reply by charbonnierg** [reply]
No, I'm suggesting to enrich metadata regarding client TLS connection. That's all.

>  If we make it really easy to use the SNI name incorrectly by not clearly indicating its not to be trusted, users will treat it as trusted.

We need to be clear here, the SNI **is trustable**. There is no way for the client to use a SNI not valid according to server config. This is **no different** from cipher or TLS version (and cipher and TLS versions are present in claims). Sure it is not enough to identify WHO is the client, and if he should be authorized, but that does not make it less "trusted".

IMHO The real question here is: should `client_tls` block contain information related to the TLS connection state, or should it only contain information related to tls certificates ?

- If the first, then it makes a lot of sense to include SNI
- If the latter, then it does not make sense to include SNI, but I don't see how `cipher` or `protocol` are justified in such case. Just like `sni`, they can be chosen by the client, and do not contain enough information by themselves to identify the client.

**Reply by derekcollison** [reply]
I am ok with adding it in, noting what R.I. said that with wildcards anyone could use multiple host names etc. in a probe.

**Reply by charbonnierg** [reply]
I do understand the fact that with wildwards, anyone could use multiple hosts. As well as I understand that certificates using alternative subject names will expose the valid set of domain names. And in any case, I only had difficulties with the concept of "trust" described by R.I. Aside from that I agree with what was said above :). Regarding the concept of trust, my understanding is as follow:

What we cannot trust is:

> the user was authorized to connect because he provided this domain name.

I understand that R.I. fears that users will jump to this conclusion and open security holes. I'm not convinced by this argument, as users of auth callout should be advanced users anyway, but I do recognize that this MAY happen...

But regardless from usage, we can definitely trust:

> the user attempted to connect using this domain name.

Which IMO, is not so far from (for example):

> the user attempted to connect using TLS protocol `1.3`

Or

> the user attempted to connect using TLS encryption cipher `TLS_AES_128_GCM_SHA256`

In the three cases, the information is metadata describing the TLS connection state of client connection.

**Reply by ripienaar** [reply]
2 of the 3 you can trust because the TLS system validated the entire TLS version or cipher mathematically. 

1 you cannot trust because any user can supply any of the valid options. 

As you pointed out the spec requires you to use SNI to then follow on to securely validate TLS handshake - this means the security lies in the validation passing. We can trust webservers to implement this requirement and not land us in hot water. 

We have no control over how this SNI name is used in callout logic and users may well not combine it with strict authentication validation (like your example does). They may think this is information that’s equally valid as “user used TLS version X wirh Y cipher”. You cannot spoof a version of TLS used or cipher used. You cannot claim to use a strong cipher when you are actually using a weak cipher. 

You can claim to connect to a name you can’t actually access and the server has no way to validate that simply by passing an invalid SNI name. 

Anyway I am done with this. Do what yous want.

**Reply by charbonnierg** [reply]
Your points are clear, although I do not completely agree with them.

Regardless of the outcome of the discussion, I'd like to express my thanks again for your time, and for this awesome project in general.
